include/orc/string_pool.hpp

```diff
@@ -16,40 +16,81 @@
 
 struct pool_string;
 
+/*
+    Stores interned strings. Thread safe in that the pool resources are per thread. 
+
+    A string pool per thread reduces the total memory usage from 83GB to 53GB. It also
+    significantly improves performance. (This is the result for the application as a whole. That
+    a per-thread memory pool works so well is perhaps counter-intuitive.)
+*/
 pool_string empool(std::string_view src);
 
+/*
+    A pool_string is an interned string. Once created, the pointer `_data` is immutable for the
+    life of the application. This has some useful properties.
+
+    * A `pool_string` is one pointer in size
+    * A `pool_string` is thread safe
+    * Two `pool_strings` pointing to the same `_data` are always equal.
+    * Becaues `pool_strings` can be in different pools, if the `_data` members are different, they may still be equal
+    * `_data` is a char* to null terminated data, which is just a c string. Useful for debugging. 
+    * if `_data` is null, it is intepreted as an empty string (""). If `_data` is not-null, it always is size() > 0
+    
+    When empooled, the hash (64 bits) and size/length (32 bits) are stored before the char* to the data.
+    Note there is no memory alignment in the pool - it is fully packed - so data needs to be memcpy'd in
+    and out, just in case the processor doesn't like un-aligned reads.
+*/
 struct pool_string {
-    pool_string() = default;
+    pool_string() {}
+    ~pool_string() {}
 
-    auto empty() const { return _s.empty(); }
+    bool empty() const { return _data == nullptr; }
 
-    std::string allocate_string() const { return std::string(_s); }
-    std::filesystem::path allocate_path() const { return std::filesystem::path(_s); }
+    std::string_view view() const {
+        if (!_data) return default_view;
+        std::uint32_t size = get_size(_data);
+        return std::string_view(_data, size);
+    }
+    
+    std::string allocate_string() const { 
+        assert(_data);
+        return std::string(view()); 
+    }
+    
+    std::filesystem::path allocate_path() const { 
+        assert(_data); 
```

Comment:
This assert isn't necessary anymore - a `pool_string` with no data can still create a valid `std::filesystem::path`.